include/ignition/utils/ImplPtr.hh

```diff
@@ -0,0 +1,251 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef IGNITION_UTILS__IMPLPTR_HH_
+#define IGNITION_UTILS__IMPLPTR_HH_
+
+#include <memory>
+#include <utility>
+
+#include <ignition/utils/detail/DefaultOps.hh>
+#include <ignition/utils/SuppressWarning.hh>
+#include <ignition/utils/Export.hh>
+
+namespace ignition
+{
+  namespace utils
+  {
+    //////////////////////////////////////////////////
+    /// \brief The ImplPtr class provides a convenient away to achieve the
+    /// <a href="http://en.cppreference.com/w/cpp/language/rule_of_three">
+    /// Rule of Zero</a> while keeping all the benefits of PIMPL. This saves us
+    /// from writing an enormous amount of boilerplate code for each class.
+    ///
+    /// To follow PIMPL design principles, create an object of this type as the
+    /// one and only member variable of your class, e.g.:
+    ///
+    /// \code
+    /// class MyClass
+    /// {
+    ///   public: /* ... public member functions ... */
+    ///
+    ///   private: class Implementation;
+    ///   private: ImplPtr<Implementation> dataPtr;
+    /// };
+    /// \endcode
+    ///
+    /// When constructing the \code{dataPtr} object, pass it
+    /// \code{MakeImpl<Implementation>(/* ... args ... */)} in the
+    /// initialization list of your class. \sa MakeImpl<T>()
+    ///
+    /// This class was inspired by the following blog post:
+    /// http://oliora.github.io/2015/12/29/pimpl-and-rule-of-zero.html
+    ///
+    /// For interface classes that should not be copiable, see the UniqueImplPtr
+    /// class further down in this header.
+    ///
+    /// \note Switching between ImplPtr and UniqueImplPtr is \em NOT ABI-safe.
+    /// This is essentially the same as changing whether or not the class
+    /// provides a copy-constructor and a copy-assignment operator, which is
+    /// bound to result in runtime linking issues at a minimum (but more
+    /// importantly, it changes the binary footprint of the class). If it is not
+    /// obvious whether a class should be copiable, then the safest choice is to
+    /// use a UniqueImplPtr and then manually add the copy constructor/operator
+    /// later if it is deemed acceptable. The next time an ABI update is
+    /// permitted, those manually written functions can be removed and the
+    /// UniqueImplPtr can be replaced with an ImplPtr.
+    template <class T,
+              class Deleter = void (*)(T*),
+              class Operations = detail::CopyMoveDeleteOperations<T> >
+    class ImplPtr
+    {
+      /// \brief Constructor
+      /// \tparam U A type that is compatible with T, i.e. either T or a class
```

Comment:
I'm not sure why we need template parameters for this constructor. And if we do need/want them, don't we need to provide templates for the copy/move ctors and assignment operators? The [SPIMPL](https://github.com/oliora/samples/blob/master/spimpl.h) implementation has does provide templates for these.

Reply:
It's been a few years since I did this, so I can't say for sure if this was the rationale, but I think the templates are used in this constructor to help make the type casting during construction a little more flexible for various types of "Deleters" and "Operations".

I think the SPIMPL implementation also includes that flexibility in its copy/move constructors and assignment operators so that it can be compatible with the broadest possible set of use cases, including the ability to cast between different SPIMPL types. I didn't bother copying over all that template machinery because I think the ignition use cases are pretty narrow. We'll never need to copy/move between `ImplPtr`s of different types.

If it's decided that those flexible copy/move operations are needed in the future, I believe it should be backwards compatible to add them later. Until then, I don't see a point in moving over all that template machinery.

Reply:
Sounds good. Thanks for the explanation. I think having the template parameters make it easy to cast between const and non-const versions and allow inheritance in `Implementation`s. But if we ever do want inheritance, I think we have to add the template parameters to the copy/move ctors as well. I haven't seen that in our codebase, but I think it's common in the Bridge design pattern.